src/network/taraxa_capability.cpp

```diff
@@ -400,17 +425,24 @@ void TaraxaCapability::interpretCapabilityPacketImpl(NodeID const &_nodeID, unsi
       break;
     }
     case GetBlocksPacket: {
-      LOG(log_dg_dag_sync_) << "Received GetBlocksPacket with " << _r.itemCount() << " known blocks";
-      std::unordered_set<blk_hash_t> known_non_finalized_blocks;
-      for (auto hash : _r) {
-        known_non_finalized_blocks.insert(blk_hash_t(hash));
-      }
+      std::unordered_set<blk_hash_t> blocks_hashes;
       std::vector<std::shared_ptr<DagBlock>> dag_blocks;
+      auto mode = static_cast<GetBlocksPacketRequestType>(_r[0].toInt<unsigned>());
+
+      if (mode == MissingHashes)
+        LOG(log_dg_dag_sync_) << "Received GetBlocksPacket with " << _r.itemCount() - 1 << " missing blocks";
+      else
+        LOG(log_dg_dag_sync_) << "Received GetBlocksPacket with " << _r.itemCount() - 1 << " known blocks";
+
+      for (size_t i = 1; i < _r.itemCount(); ++i) {
+        blocks_hashes.emplace(_r[i]);
+      }
+
       const auto &blocks = dag_mgr_->getNonFinalizedBlocks();
       for (auto &level_blocks : blocks) {
         for (auto &block : level_blocks.second) {
           auto hash = blk_hash_t(block);
-          if (known_non_finalized_blocks.count(hash) == 0) {
+          if (blocks_hashes.count(hash) == (mode == MissingHashes)) {
```

Comment:
I am just wandering is there a possibility we get into some never ending loop of requesting blocks and not receiving them. I mean in theory it should not happen but if for some reason we have a bug where a node is missing something it should not be missing, we might flood the networking with a loop of getBlocksPacket and BlocksPacket. I think you should check here if the node has the missinghashes blocks. If it does not have them, I would just log an error and not send back the BlocksPacket to break such a loop

Reply:
I think this will not happened, because I am not processing block with missing tip/pivot, so if you do not send me what I need I will not re-request it again. If I add check there you wanted, what will happened if you send me blocks in wrong order and I needed pivot is last one, then I can to ask you again to send it to me because you will mark I have them

Reply:
But I thought we request for the block that had a missing tip/pivot as well since we drop it in first attempt. If we do not do that how do we ever get the original block again

Reply:
Let's discuss this over tomorrow's call